index/scorch/snapshot_index.go

```diff
@@ -492,109 +491,64 @@ func (i *IndexSnapshot) DocumentVisitFieldTerms(id index.IndexInternalID,
 }
 
 func (i *IndexSnapshot) documentVisitFieldTerms(id index.IndexInternalID,
-	fields []string, visitor index.DocumentFieldTermVisitor, dvs segment.DocVisitState) (
-	segment.DocVisitState, error) {
-
+	fields []string, visitor index.DocumentFieldTermVisitor,
+	dvs segment.DocVisitState) (segment.DocVisitState, error) {
 	docNum, err := docInternalToNumber(id)
 	if err != nil {
 		return nil, err
 	}
+
 	segmentIndex, localDocNum := i.segmentIndexAndLocalDocNumFromGlobal(docNum)
 	if segmentIndex >= len(i.segment) {
 		return nil, nil
 	}
 
 	ss := i.segment[segmentIndex]
 
-	if zaps, ok := ss.segment.(segment.DocumentFieldTermVisitable); ok {
-		// get the list of doc value persisted fields
-		pFields, err := zaps.VisitableDocValueFields()
+	var vFields []string // fields that are visitable via the segment
+
+	ssv, ssvOk := ss.segment.(segment.DocumentFieldTermVisitable)
+	if ssvOk && ssv != nil {
+		vFields, err = ssv.VisitableDocValueFields()
 		if err != nil {
 			return nil, err
 		}
-		// assort the fields for which terms look up have to
-		// be performed runtime
-		dvPendingFields := extractDvPendingFields(fields, pFields)
-		// all fields are doc value persisted
-		if len(dvPendingFields) == 0 {
-			return zaps.VisitDocumentFieldTerms(localDocNum, fields, visitor, dvs)
-		}
+	}
 
-		// concurrently trigger the runtime doc value preparations for
-		// pending fields as well as the visit of the persisted doc values
-		errCh := make(chan error, 1)
+	// cFields represents the fields that we'll need from the cachedDocs
+	cFields := subtractStrings(fields, vFields)
+
+	var errCh chan error
+
+	if !ss.cachedDocs.hasFields(cFields) {
```

Comment:
As an other optimization here, hasFields can be renamed to fieldsToPrepare and this could return only the set of fields that aren't in cachedDocs already, and we invoke prepareFields only for that subset rather than all of cFields?

Reply:
another good one!

Reply:
actually, on more thought, computing the subset means another allocation of some slice or something to hold the "subset", and prepareFields() looks efficient enough to me for this case